pm: device: Several fixes

ceolin opened the pull request:
- Don't wake up devices unnecessarily
 - Fix concurrence issues with `k_poll_signal`. Using `k_condvar` instead.
 - Fix pre-kernel API usage
 - Some code guideline fixes
 - Use spin lock instead of semaphore to protect critical sessions, since it was the semaphore with K_FOREVER which is not allowed from ISR.
 - Change API signature for clarity sake

Commit:
power: device: Don't wake up devices unnecessarily

Currently when the system goes to sleep it asks *all* devices that
support PM to suspend or go to a low power state, and the system wakes
up it put *all* suspended devices in active state, even if a device
was already suspended and not being used.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: device: void *context -> uint32_t *state

The context parameter used across device power management is
actually the power state. Just use it and avoid a lot of
unnecessary casts.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: device_pm: Check if is running in pre-kernel

Don't use kernel services before in pre-kernel state.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: device_pm: Use BIT macro for code compliance

BIT macro properly handle bit shift operations.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: device_pm: Fix concurrence issues

The sync API was using k_poll_signal and in certain conditions is
possible multiple threads waiting on a signal leading to an undefined
behavior.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: device_pm: Add missing fallthrough

Add an explicit fallthrough in a case label.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: device_pm: Use spin lock instead of semaphore

Device pm runtime was using semaphore to protect critical section but
enable / disable functions were waiting on the semaphore. So, just
replace it with a spin lock.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

dcpleung commented on include/device.h:
```diff
@@ -333,7 +333,7 @@ typedef int16_t device_handle_t;
 
 typedef void (*device_pm_cb)(const struct device *dev,
-			     int status, void *context, void *arg);
+			     int status, uint32_t *state, void *arg);
```
Maybe use enum instead?

ceolin replied:
I intend to do it, but I will need to review the states and unify with fsm_state. There is another pr doing a big cleanup and will be better to have that one before doing this type change. Here is just change the signature since all places were using it was doing cast o uint32_t

gmarull commented:
What is the advantage of using spinlocks here? Can you explain in some more detail?

gmarull reviewed (commented):
Added one question

ceolin commented:
> What is the advantage of using spinlocks here? Can you explain in some more detail?

You cannot use K_FOREVER with semaphores form an ISR.

sjg20 commented on subsys/power/device_pm.c:
```diff
@@ -122,25 +126,19 @@ static int device_pm_request(const struct device *dev,
+	k_mutex_unlock(&request_mutex);
 
-	return result == target_state ? 0 : -EIO;
+	return target_state == atomic_get(&dev->pm->fsm_state) ? 0 : -EIO;
```
This code is a bit hard to understand. How does it actually end up being in the target state? How does it know what to do, since target_state is not stored anywhere.

ceolin replied:
you request to device to change to `target_state`, as this operation may be asynchronous the device is responsible to call the given callback with its state after this operation. The state given by the device is then stored in `dev->pm->fsm_state` we then compare here to check if the state was properly set.

ceolin commented on drivers/display/display_st7789v.c:
```diff
@@ -409,14 +409,14 @@ static void st7789v_enter_sleep(struct st7789v_data *data)
 
 static int st7789v_pm_control(const struct device *dev, uint32_t ctrl_command,
-				 void *context, device_pm_cb cb, void *arg)
+				 uint32_t *state, device_pm_cb cb, void *arg)
```
the last parameter is supposed to be used as context to the callback. This variable that was being called `void *context` was being treated as an `uint32_t *` by the device_pm framework (and everywhere else), so it is not like you could just change it to be a different type in your use case. If you need to pass a different data you should use `*arg`. 

The reason to change it is, as you can see, very confuse and with bad variable names this function signature.

pfalcon replied:
Still unclear why it wasn't made into `struct pm_context { uint32_t state };` to allow for seamless extensibility going forward, but ok...

ceolin commented on subsys/power/device.c:
```diff
@@ -60,26 +60,63 @@ static device_idx_t num_pm;
 /* Number of devices successfully suspended. */
 static device_idx_t num_susp;
 
+static bool should_suspend(const struct device *dev, uint32_t state)
```
yep:
```
./samples/drivers/spi_flash_at45/prj.conf:CONFIG_PM_DEVICE=y
./samples/subsys/power/device_pm/prj.conf:CONFIG_PM_DEVICE=y
./samples/boards/ti/cc13x2_cc26x2/system_off/prj.conf:CONFIG_PM_DEVICE=y
./samples/boards/nrf/system_off/prj.conf:CONFIG_PM_DEVICE=y
./samples/boards/stm32/power_mgmt/blinky/prj.conf:CONFIG_PM_DEVICE=y
./tests/net/pm/prj.conf:CONFIG_PM_DEVICE=y
./tests/subsys/power/power_mgmt/prj.conf:CONFIG_PM_DEVICE=y
```

ceolin commented:
@sjg20 is not clear for me what changes were requested, there is anything other than the typo that needs to be changed after my answers ?

sjg20 reviewed (approved):
Looks good. I do think a comment on the operation (like you provided in the response) would be useful

ceolin commented:
> Looks good. I do think a comment on the operation (like you provided in the response) would be useful

will add it, thanks for the feedback

Commit:
power: device: Do not suspend a busy device

Check if a device is busy before suspend it.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device: fsm_state -> state

Rename state variable.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

erwango commented on subsys/pm/device_runtime.c:
```diff
@@ -111,6 +111,16 @@ static int pm_device_request(const struct device *dev,
 		}
 	}
 
+	if (k_is_pre_kernel()) {
```
I'm not convinced this is the correct way.
This means clock activation should be done on device side, on top of calling pm_device_get primitive (which at least, should be documented).
Instead, isn't it a way to run the worker routine code in a synchronous way at this stage ?

ceolin replied:
This exactly code was part of my other pr
https://github.com/zephyrproject-rtos/zephyr/pull/33942/commits/64dac719dd5762dc46b5cb3e7d1774091feaf1ed

The reason for what I have not included it in this one is because this is a workaround from my understanding. At this point devices should have being initialized and powered on accordingly and this activation unnecessary. This is fixing an existing problem not in the proper way imho.

But as we don't have it fixed, I can cherry-pick this commit and once we have it done properly I remove this path.

ceolin replied:
Btw, this is not about concurrence, but putting the device in a unexpected condition. From the device perspective is expected the init function be called before the power control function. The device may do some data initialization in the init that will be used in the power control  ....

nashif replied:
> At this point devices should have being initialized and powered on accordingly and this activation unnecessary.

No, this is not the case. you need to be able to use those drivers and use them means powering them up if they are not active. The issue was originally exactly that, we are trying to power on a device and cant be cause kernel is still not up, i.e. we try to deal with code designed for multithreds and concurrency that in this stage can be skipped.

ceolin replied:
yep, but not in pre-kernel. Honestly I don't see why devices are not initialized in the right order at this point. Anyway, I've cherry-picked that commit

nashif replied:
what does initialized mean exactly? init() AFAIK does not power up a device necessarily..

ceolin replied:
call init()

Commit:
pm: device_runtime: Power device on/off in pre-kernel

Devices may be initialized but started powered down for this reason
is necessary to power a device on if requested even if in pre-kernel
state.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>
